File: tests/fixtures/polyglot_baseline/go_backend/services/user_service.go
package services

import (
	"context"
	"fmt"
	"strings"

	"backend/models"
	"backend/repository"
)

type UserService struct {
	repo repository.UserRepository
}

func NewUserService(repo repository.UserRepository) *UserService {
	return &UserService{repo: repo}
}

func (s *UserService) GetAllUsers(ctx context.Context) ([]*models.User, error) {
	users, err := s.repo.GetAll(ctx)
	if err != nil {
		return nil, fmt.Errorf("failed to get all users: %w", err)
	}
	return users, nil
}

func (s *UserService) GetUserByID(ctx context.Context, id int64) (*models.User, error) {
	user, err := s.repo.GetByID(ctx, id)
	if err != nil {
		if strings.Contains(err.Error(), "not found") {
			return nil, models.ErrUserNotFound
		}
		return nil, fmt.Errorf("failed to get user by id: %w", err)
	}
	return user, nil
}

func (s *UserService) GetUserByEmail(ctx context.Context, email string) (*models.User, error) {
	if email == "" {
		return nil, fmt.Errorf("email cannot be empty")
	}

	user, err := s.repo.GetByEmail(ctx, email)
	if err != nil {
		if strings.Contains(err.Error(), "not found") {
			return nil, models.ErrUserNotFound
		}
		return nil, fmt.Errorf("failed to get user by email: %w", err)
	}
	return user, nil
}

func (s *UserService) CreateUser(ctx context.Context, user *models.User) (*models.User, error) {
	if user.Email == "" || user.Name == "" {
		return nil, fmt.Errorf("email and name are required")
	}

	existing, err := s.repo.GetByEmail(ctx, user.Email)
	if err == nil && existing != nil {
		return nil, models.ErrUserAlreadyExists
	}

	created, err := s.repo.Create(ctx, user)
	if err != nil {
		return nil, fmt.Errorf("failed to create user: %w", err)
	}

	return created, nil
}

func (s *UserService) UpdateUser(ctx context.Context, id int64, name string) (*models.User, error) {
	if name == "" {
		return nil, fmt.Errorf("name cannot be empty")
	}

	existing, err := s.repo.GetByID(ctx, id)
	if err != nil {
		return nil, models.ErrUserNotFound
	}

	existing.Name = name
	updated, err := s.repo.Update(ctx, existing)
	if err != nil {
		return nil, fmt.Errorf("failed to update user: %w", err)
	}

	return updated, nil
}

func (s *UserService) DeleteUser(ctx context.Context, id int64) error {
	user, err := s.repo.GetByID(ctx, id)
	if err != nil {
		return models.ErrUserNotFound
	}

	if err := s.repo.Delete(ctx, user); err != nil {
		return fmt.Errorf("failed to delete user: %w", err)
	}

	return nil
}

File: tests/fixtures/polyglot_baseline/python_service/services/user_service.py
"""
User business logic service.

Handles user CRUD operations, validation, and business rules.
"""

from typing import Any, Optional

from ..exceptions import ConflictError, NotFoundError, ValidationError
from ..models.permissions import RoleEnum
from ..utils import hash_password
from ..utils.validators import validate_email, validate_username


class UserService:
    """Service for user operations."""

    def __init__(self, db: "Database"):  # noqa: F821
        """Initialize user service with database dependency."""
        self.db = db
        # In-memory store for demonstration (replace with real DB in production)
        self.users: dict[int, dict[str, Any]] = {}
        self.next_id = 1

    def create_user(
        self,
        email: str,
        username: str,
        password: str,
        full_name: Optional[str] = None,
        role: RoleEnum = RoleEnum.MEMBER,
    ) -> dict:
        """
        Create a new user.

        Args:
            email: User email (must be unique)
            username: Username (must be unique)
            password: Plain text password
            full_name: Optional full name
            role: User role

        Returns:
            Created user dictionary

        Raises:
            ValidationError: If validation fails
            ConflictError: If email or username already exists
        """
        email = validate_email(email)
        username = validate_username(username)

        # Check for duplicates
        if any(u["email"] == email for u in self.users.values()):
            raise ConflictError(f"Email already registered: {email}")

        if any(u["username"] == username for u in self.users.values()):
            raise ConflictError(f"Username already taken: {username}")

        user_id = self.next_id
        self.next_id += 1

        user = {
            "id": user_id,
            "email": email,
            "username": username,
            "password_hash": hash_password(password),
            "full_name": full_name,
            "role": role,
            "is_active": True,
            "created_at": "2024-01-01T00:00:00",
            "updated_at": "2024-01-01T00:00:00",
        }

        self.users[user_id] = user
        return {k: v for k, v in user.items() if k != "password_hash"}

    def get_user_by_id(self, user_id: int) -> dict:
        """
        Get user by ID.

        Args:
            user_id: User ID

        Returns:
            User dictionary

        Raises:
            NotFoundError: If user not found
        """
        if user_id not in self.users:
            raise NotFoundError("User", str(user_id))
        return self.users[user_id]

    def get_user_by_email(self, email: str) -> dict:
        """
        Get user by email.

        Args:
            email: User email

        Returns:
            User dictionary

        Raises:
            NotFoundError: If user not found
        """
        email = validate_email(email)
        for user in self.users.values():
            if user["email"] == email:
                return user
        raise NotFoundError("User", email)

    def list_users(self, skip: int = 0, limit: int = 10) -> tuple[list[dict], int]:
        """
        List all users with pagination.

        Args:
            skip: Number of users to skip
            limit: Maximum users to return

        Returns:
            Tuple of (user list, total count)
        """
        users_list = list(self.users.values())
        total = len(users_list)
        users_list = users_list[skip : skip + limit]
        return users_list, total

    def update_user(self, user_id: int, **updates: Any) -> dict:
        """
        Update user information.

        Args:
            user_id: User ID
            **updates: Fields to update

        Returns:
            Updated user dictionary

        Raises:
            NotFoundError: If user not found
            ValidationError: If validation fails
        """
        user = self.get_user_by_id(user_id)

        if "email" in updates:
            updates["email"] = validate_email(updates["email"])

        if "username" in updates:
            updates["username"] = validate_username(updates["username"])

        user.update(updates)
        return {k: v for k, v in user.items() if k != "password_hash"}

    def delete_user(self, user_id: int) -> bool:
        """
        Delete user.

        Args:
            user_id: User ID to delete

        Returns:
            True if deleted

        Raises:
            NotFoundError: If user not found
        """
        if user_id not in self.users:
            raise NotFoundError("User", str(user_id))

        del self.users[user_id]
        return True

    def update_password(self, user_id: int, password_hash: str) -> bool:
        """
        Update user password hash.

        Args:
            user_id: User ID
            password_hash: New password hash

        Returns:
            True if updated
        """
        user = self.get_user_by_id(user_id)
        user["password_hash"] = password_hash
        return True

    def deactivate_user(self, user_id: int) -> bool:
        """Deactivate user account."""
        user = self.get_user_by_id(user_id)
        user["is_active"] = False
        return True

    def activate_user(self, user_id: int) -> bool:
        """Activate user account."""
        user = self.get_user_by_id(user_id)
        user["is_active"] = True
        return True
